Pulling email from upstream idp token if missing

MatthewWilkes opened the pull request:
This changes the way that the email address is determined to progressively look in different places:

1. The email claim
2. The first entry of the emails claim
3. The email claim of the upstream idp token
4. The upn claim of the upstream idp token

This change does not validate the signing of the upstream IdP's JWT as we do not have any credentials to do so. Those are kept in B2C. However, the claim as a whole is part of the signature from the B2C instance that we trust, so we transitively trust that it's valid as the B2C instance will have validated.

MatthewWilkes committed:
Initial work to pulling email from upstream idp token if missing

Polynya commented on openid_connect_azure_b2c.module:
```diff
@@ -1,8 +1,21 @@
 use Drupal\openid_connect_azure_b2c\Plugin\OpenIDConnectClient\OpenIDConnectB2CClient;
 
+
+function parseToken(string $token) {
```
Can the return type be declared e.g.
`function parseToken(string $token): array {`
If this is done then the fallback could be `return [];`

MatthewWilkes replied:
I pulled this from the core openid_connect module, so left it untouched. Could do, though?

MatthewWilkes replied:
Updated in 73a6c4f

Polynya replied:
Ah ok, you have to be careful if you're overriding a parent method.

MatthewWilkes replied:
It's not a parent, otherwise I'd have called it directly. It's part of a support view for a different part of the system, but it's protected so not easily accessed.

MatthewWilkes committed:
Refactor email extraction to be part of the plugin

This moves the email extraction logic into the plugin proper and unit tests it.

Polynya commented on src/Plugin/OpenIDConnectClient/OpenIDConnectB2CClient.php:
```diff
@@ -75,16 +74,60 @@ public function submitConfigurationForm(array &$form, FormStateInterface $form_s
\ No newline at end of file
+  /**
+   * Parse the token from upstream.
+   */
```
Not sure why CodeSniffer didn't spot this missing for doc
```
@param string $token
  Token.
```

MatthewWilkes replied:
Fixed this and the other instance.

MatthewWilkes committed:
Address extract comment

Update the extract token method to return only valid token contents, and mark the return type.

Polynya commented on tests/src/Unit/ExtractEmailTest.php:
```diff
@@ -0,0 +1,104 @@
+  }
+
+  /**
+   *
```
Comment at top of each method

MatthewWilkes replied:
Awww, fiiine.

MatthewWilkes replied:
Added to all.

MatthewWilkes committed:
Docstring and codesmell fixes

MatthewWilkes commented on openid_connect_azure_b2c.module:
```diff
@@ -13,8 +13,7 @@ function openid_connect_azure_b2c_openid_connect_userinfo_alter(array &$userinfo
-  // Use the 0th email as the main one
-  $userinfo['email'] = $userinfo['emails'][0];
+  
+  $userinfo['email'] = $plugin->extractEmail($userinfo);
```
@Polynya Can I do this now that it's a static method, or should I call `OpenIDConnectB2CClient::extractEmail`?

MatthewWilkes replied:
Looks like it works :)
